site/content/blog/2021-04-23-optimizing-go-compiler-2.md

```diff
@@ -0,0 +1,343 @@
+---
+title: "Fixing a bug in the Go compiler as a newbie: a deep dive (II)"
+heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (II)"
+description: ""
+summary: "The second part of this series of posts details the issue we are trying to fix. We are using one of the tools we learned in the first part, Static Single Assignment, to understand each and every line of the intermediate code generated by the compiler. We will even learn a tiny bit of assembly in the way."
+slug: optimizing-go-compiler-2
+date: 2021-04-23T12:01:00-05:00
+categories:
+    - "go"
+author: Alejandro García Montoro
+github: agarciamontoro
+community: alejandro.garcia
+---
+
+This is the second part of a series of posts on the compiler of Go. Make sure you have read [the first part](/blog/optimizing-go-compiler-1) before going on.
+
+In the first part of the series we learned what the issue is and the role of SSA and the rewrite rules in the Go compiler. Let's try and use what we know now to actually investigate the bug.
+
+# The investigation
+
+As we learned in the first part, the issue we need to investigate is that the generated code for the following function is not as optimized as it could be:
+
+```go
+import "encoding/binary"
+
+func copyArrayToSlice(dest []byte, src *[8]byte) {
+	_ = dest[8]                                // 1. bounds check
+	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
+	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
```

Comment:
destiny slice is a really cool slice. Let's fix it ;)

Reply:
Oh man, we use the same word in Spanish for both destiny and destination :joy: Fixed in https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/bcd613fd20785b2de630c486a25b70caa5dfc2fe, thanks!